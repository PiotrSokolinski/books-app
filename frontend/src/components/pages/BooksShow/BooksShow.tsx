import { Show, useTranslate, EditButton } from 'react-admin';
import { useParams } from 'react-router-dom';

import { BookActions } from '../../organisms/BookActions/BookActions';
import { BookCard } from '../../organisms/BookCard/BookCard';
import { Wrapper } from '../../templates/Wrapper/Wrapper';

export const BooksShow = () => {
  const translate = useTranslate();
  const { id } = useParams();

  return (
    <Wrapper>
      <Show
        id={id}
        title={translate('booksShow.barTitle')}
        actions={<BookActions action={<EditButton />} />}
      >
        <BookCard />
      </Show>
    </Wrapper>
  );
};
