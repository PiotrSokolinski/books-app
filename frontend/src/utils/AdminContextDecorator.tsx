import React from 'react';
import {
  AdminContext,
  DataProvider,
  GetListResult,
  memoryStore,
  testDataProvider,
} from 'react-admin';

import { theme } from '../theme/theme';
import { i18nProvider } from '../i18n/i18n';
import { Book } from '../types/book.ts';

const defaultDataProvider: DataProvider = testDataProvider({
  getList: async (): Promise<GetListResult<Book>> => {
    const books: Book[] = [
      {
        id: '1',
        title: 'The Hobbit',
        author: 'J.R.R. Tolkien',
        isbnNumber: '1234567890',
        numberOfPages: 300,
        rating: 4,
      },
      {
        id: '2',
        title: 'The Lord of the Rings',
        author: 'J.R.R. Tolkien',
        isbnNumber: '0987654321',
        numberOfPages: 1000,
        rating: 5,
      },
    ];

    return Promise.resolve({
      total: books.length,
      data: books,
    });
  },
  getOne: () => {
    return Promise.resolve({
      data: {
        id: '1',
        title: 'The Lord of the Rings',
        author: 'J.R.R. Tolkien',
        isbnNumber: '0987654321',
        numberOfPages: 1000,
        rating: 5,
      },
    });
  },
  create: () => {
    return Promise.resolve({
      data: {
        id: '3',
        title: 'Test Book Title',
        author: 'J.R.R. Tolkien',
        isbnNumber: '0987654321',
        numberOfPages: 1000,
        rating: 5,
      },
    });
  },
  update: () => {
    return Promise.resolve({
      data: {
        id: '2',
        title: 'Test Book Title',
        author: 'J.R.R. Tolkien',
        isbnNumber: '0987654321',
        numberOfPages: 1000,
        rating: 5,
      },
    });
  },
  delete: () => {
    return Promise.resolve({
      data: {
        id: '3',
        title: 'Test Book Title',
        author: 'J.R.R. Tolkien',
        isbnNumber: '0987654321',
        numberOfPages: 1000,
        rating: 5,
      },
    });
  },
} as unknown as DataProvider); // https://github.com/marmelab/react-admin/issues/5476

// TODO: F7 - Add more documentation and comments to the codebase
/**
 * A decorator component that provides the React-Admin context for testing or storybook environments.
 *
 * This component wraps its children with the necessary React-Admin context providers,
 * including a theme, internationalization provider, and optionally a `dataProvider`.
 * The `authProvider` is set to `undefined`, as this provider is designed for simple use cases.
 *
 * It supports passing a custom `dataProvider` for testing or storybook purposes.
 *
 * @param {Object} props - The props for the component.
 * @param {React.ReactNode} props.children - The child components to render within the React-Admin context.
 * @param {Partial<DataProvider>} [props.dataProvider] - An optional custom `dataProvider` to be used in place of the default one.
 *
 * @returns {JSX.Element} The wrapped children within the React-Admin context, with context providers like theme and internationalization applied.
 *
 * @example
 * <AdminContextDecorator>
 *   <MyComponent />
 * </AdminContextDecorator>
 */
export const AdminContextDecorator: React.FC<{
  children: React.ReactNode;
  dataProvider?: Partial<DataProvider>;
}> = ({
  children,
  dataProvider,
}: {
  children: React.ReactNode;
  dataProvider?: Partial<DataProvider>;
}): React.JSX.Element => (
  <AdminContext
    dataProvider={
      dataProvider ? testDataProvider(dataProvider) : defaultDataProvider
    }
    authProvider={undefined}
    theme={theme}
    store={memoryStore()}
    i18nProvider={i18nProvider}
  >
    {children}
  </AdminContext>
);
