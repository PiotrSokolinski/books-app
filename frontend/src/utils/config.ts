const { VITE_API_URL } = import.meta.env;

export const APP_CONFIG = {
  VITE_API_URL,
};
