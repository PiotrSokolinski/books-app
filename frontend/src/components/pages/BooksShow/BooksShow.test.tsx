import { ResourceContextProvider } from 'react-admin';
import { fireEvent } from '@testing-library/react';
import { render, screen, waitFor } from '@testing-library/react';
import { vi, describe, it, expect } from 'vitest';

import { BooksShow } from './BooksShow';
import { AdminContextDecorator } from '../../../utils/AdminContextDecorator';

vi.mock(import('react-admin'), async (importOriginal) => {
  const actual = await importOriginal();
  return {
    ...actual,
    useResourceContext: vi.fn().mockReturnValue('books'),
    useShowController: vi.fn().mockReturnValue({
      record: {
        id: '1',
        title: 'Test Book',
        author: 'Test Author',
        isbnNumber: '1234567890',
        numberOfPages: 200,
        rating: 5,
      },
      loading: false,
      error: null,
    }),
  };
});

vi.mock(import('react-router-dom'), async (importOriginal) => {
  const actual = await importOriginal();

  return {
    ...actual,
    useParams: vi.fn().mockReturnValue({ id: '1' }),
  };
});

describe('BooksShow', () => {
  it('renders the BooksShow component and click edit', async () => {
    render(
      <AdminContextDecorator>
        <ResourceContextProvider value="books">
          <BooksShow />
        </ResourceContextProvider>
      </AdminContextDecorator>,
    );

    await waitFor(() => {
      const showButton = screen.getByText(/edit/i);
      expect(showButton).toBeInTheDocument();
      fireEvent.click(showButton);
    });
  });

  it('renders the BooksShow component and go back to list', async () => {
    render(
      <AdminContextDecorator>
        <ResourceContextProvider value="books">
          <BooksShow />
        </ResourceContextProvider>
      </AdminContextDecorator>,
    );

    await waitFor(() => {
      const backButton = screen.getByText('Back to list');
      expect(backButton).toBeInTheDocument();
      fireEvent.click(backButton);
    });
  });

  it('matches the snapshot', async () => {
    const { asFragment } = render(
      <AdminContextDecorator>
        <ResourceContextProvider value="books">
          <BooksShow />
        </ResourceContextProvider>
      </AdminContextDecorator>,
    );
    await waitFor(() => {
      const showButton = screen.getByText(/edit/i);
      expect(showButton).toBeInTheDocument();
    });
    expect(asFragment()).toMatchSnapshot();
  });
});
