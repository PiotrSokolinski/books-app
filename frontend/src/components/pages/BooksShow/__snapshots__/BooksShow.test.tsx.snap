// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`BooksShow > matches the snapshot 1`] = `
<DocumentFragment>
  <div
    class="MuiBox-root css-1es8s0n"
  >
    <div
      class="show-page css-cg7rje-RaShow-root"
    >
      <div
        class="MuiToolbar-root MuiToolbar-dense css-8czsxw-MuiToolbar-root-RaTopToolbar-root"
      >
        <div
          class="MuiBox-root css-1oxffa5"
        >
          <button
            aria-label="Back to list"
            class="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeSmall MuiButton-textSizeSmall MuiButton-colorPrimary MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeSmall MuiButton-textSizeSmall MuiButton-colorPrimary css-7nabzj-MuiButtonBase-root-MuiButton-root-RaButton-root"
            tabindex="0"
            type="button"
          >
            <span
              class="MuiButton-icon MuiButton-startIcon MuiButton-iconSizeSmall css-y6rp3m-MuiButton-startIcon"
            >
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                data-testid="ArrowBackIcon"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M20 11H7.83l5.59-5.59L12 4l-8 8 8 8 1.41-1.41L7.83 13H20z"
                />
              </svg>
            </span>
            Back to list
            <span
              class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
            />
          </button>
          <a
            aria-label="Edit"
            class="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeSmall MuiButton-textSizeSmall MuiButton-colorPrimary MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeSmall MuiButton-textSizeSmall MuiButton-colorPrimary RaEditButton-root css-1pbrwyd-MuiButtonBase-root-MuiButton-root-RaButton-root-RaEditButton-root"
            href="#/books/1"
            tabindex="0"
          >
            <span
              class="MuiButton-icon MuiButton-startIcon MuiButton-iconSizeSmall css-y6rp3m-MuiButton-startIcon"
            >
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                data-testid="CreateIcon"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M3 17.25V21h3.75L17.81 9.94l-3.75-3.75zM20.71 7.04c.39-.39.39-1.02 0-1.41l-2.34-2.34a.9959.9959 0 0 0-1.41 0l-1.83 1.83 3.75 3.75z"
                />
              </svg>
            </span>
            Edit
            <span
              class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
            />
          </a>
        </div>
      </div>
      <div
        class="RaShow-main"
      >
        <div
          class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiCard-root RaShow-card css-bhp9pd-MuiPaper-root-MuiCard-root"
        >
          <div
            class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiCard-root css-bhp9pd-MuiPaper-root-MuiCard-root"
          >
            <div
              class="MuiCardContent-root css-46bh2p-MuiCardContent-root"
            >
              <span
                class="MuiTypography-root MuiTypography-h4 css-5lbw0b-MuiTypography-root"
              >
                The Lord of the Rings
              </span>
              <hr
                class="MuiDivider-root MuiDivider-fullWidth css-1b008ma-MuiDivider-root"
              />
              <div
                class="MuiBox-root css-8cxpeo"
              >
                <p
                  class="MuiTypography-root MuiTypography-body1 css-ahj2mt-MuiTypography-root"
                >
                  Author:
                </p>
                <span
                  class="MuiTypography-root MuiTypography-h6 css-snu5bf-MuiTypography-root"
                >
                  J.R.R. Tolkien
                </span>
              </div>
              <div
                class="MuiBox-root css-8cxpeo"
              >
                <p
                  class="MuiTypography-root MuiTypography-body1 css-ahj2mt-MuiTypography-root"
                >
                  Number of pages:
                </p>
                <span
                  class="MuiTypography-root MuiTypography-h6 css-snu5bf-MuiTypography-root"
                >
                  1000
                </span>
              </div>
              <div
                class="MuiBox-root css-8cxpeo"
              >
                <p
                  class="MuiTypography-root MuiTypography-body1 css-ahj2mt-MuiTypography-root"
                >
                  ISBN number:
                </p>
                <span
                  class="MuiTypography-root MuiTypography-h6 css-snu5bf-MuiTypography-root"
                >
                  0987654321
                </span>
              </div>
              <div
                class="MuiBox-root css-8cxpeo"
              >
                <p
                  class="MuiTypography-root MuiTypography-body1 css-ahj2mt-MuiTypography-root"
                >
                  Rating:
                </p>
                <div
                  class="MuiBox-root css-0"
                >
                  <svg
                    aria-hidden="true"
                    class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                    data-testid="StarIcon"
                    focusable="false"
                    viewBox="0 0 24 24"
                  >
                    <path
                      d="M12 17.27 18.18 21l-1.64-7.03L22 9.24l-7.19-.61L12 2 9.19 8.63 2 9.24l5.46 4.73L5.82 21z"
                    />
                  </svg>
                  <svg
                    aria-hidden="true"
                    class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                    data-testid="StarIcon"
                    focusable="false"
                    viewBox="0 0 24 24"
                  >
                    <path
                      d="M12 17.27 18.18 21l-1.64-7.03L22 9.24l-7.19-.61L12 2 9.19 8.63 2 9.24l5.46 4.73L5.82 21z"
                    />
                  </svg>
                  <svg
                    aria-hidden="true"
                    class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                    data-testid="StarIcon"
                    focusable="false"
                    viewBox="0 0 24 24"
                  >
                    <path
                      d="M12 17.27 18.18 21l-1.64-7.03L22 9.24l-7.19-.61L12 2 9.19 8.63 2 9.24l5.46 4.73L5.82 21z"
                    />
                  </svg>
                  <svg
                    aria-hidden="true"
                    class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                    data-testid="StarIcon"
                    focusable="false"
                    viewBox="0 0 24 24"
                  >
                    <path
                      d="M12 17.27 18.18 21l-1.64-7.03L22 9.24l-7.19-.61L12 2 9.19 8.63 2 9.24l5.46 4.73L5.82 21z"
                    />
                  </svg>
                  <svg
                    aria-hidden="true"
                    class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                    data-testid="StarIcon"
                    focusable="false"
                    viewBox="0 0 24 24"
                  >
                    <path
                      d="M12 17.27 18.18 21l-1.64-7.03L22 9.24l-7.19-.61L12 2 9.19 8.63 2 9.24l5.46 4.73L5.82 21z"
                    />
                  </svg>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
